coverage: Don't bother renumbering expressions on the Rust side

Zalathar opened the pull request:
The LLVM API that we use to encode coverage mappings already has its own code for removing unused coverage expressions and renumbering the rest.

This lets us get rid of our own complex renumbering code, making it easier to change our coverage code in other ways.

---

Now that we have tests for coverage mappings (#114843), I've been able to verify that this PR doesn't make the coverage mappings worse, thanks to an explicit simplification step.

Zalathar committed:
coverage: Don't bother renumbering expressions on the Rust side

The LLVM API that we use to encode coverage mappings already has its own code
for removing unused coverage expressions and renumbering the rest.

This lets us get rid of our own complex renumbering code, making it easier to
change our coverage code in other ways.

Zalathar committed:
coverage: Make the zero counter a constant

Zalathar commented:
@rustbot label +A-code-coverage

Swatinem commented on compiler/rustc_codegen_llvm/src/coverageinfo/map_data.rs:
```diff
@@ -146,8 +143,14 @@ impl<'tcx> FunctionCoverage<'tcx> {
+        // Expression IDs are indices into `self.expressions`, and on the LLVM
+        // side they will be treated as indices into `counter_expressions`, so
+        // the two vectors should correspond 1:1.
+        assert_eq!(self.expressions.len(), counter_expressions.len());
```
A debug assertion may be sufficient here? With the code right now, this is trivially true, as you are doing a `self.expressions.iter()...collect()`. So this is just a guard so this assumption does not change with any future refactors.

Zalathar replied:
Given that it’s trivially true, I don’t see any reason *not* to use a full assertion here, especially since it should get optimised out anyway.

(Note also that debug assertions aren’t checked in `run-coverage` tests by default.)

Zalathar commented on compiler/rustc_codegen_llvm/src/coverageinfo/map_data.rs:
```diff
@@ -163,149 +166,50 @@ impl<'tcx> FunctionCoverage<'tcx> {
-                    id_to_counter(&new_indexes, rhs).map(|rhs_counter| (lhs_counter, rhs_counter))
-                })
-            {
-                if lhs_counter.is_zero() && op.is_subtract() {
-                    // The left side of a subtraction was probably optimized out. As an example,
-                    // a branch condition might be evaluated as a constant expression, and the
-                    // branch could be removed, dropping unused counters in the process.
-                    //
-                    // Since counters are unsigned, we must assume the result of the expression
-                    // can be no more and no less than zero. An expression known to evaluate to zero
-                    // does not need to be added to the coverage map.
-                    //
-                    // Coverage test `loops_branches.rs` includes multiple variations of branches
-                    // based on constant conditional (literal `true` or `false`), and demonstrates
-                    // that the expected counts are still correct.
-                    debug!(
-                        "Expression subtracts from zero (assume unreachable): \
-                        original_index={:?}, lhs={:?}, op={:?}, rhs={:?}, region={:?}",
-                        original_index, lhs, op, rhs, optional_region,
-                    );
-                    rhs_counter = Counter::zero();
```
This is the part I mentioned not being able to trigger in my testing. In particular, the mentioned `loops_branches.rs` test doesn't appear to trigger it.

Note that even if this situation does occur for other code, coverage reports should still be correct with or without this simplification.

Zalathar replied:
With the new test suite I was able to observe cases where this simplification seems to trigger, so I was able to update the new code with similar simplifications so that the mappings don't regress.

Zalathar commented:
I'm still happy for this to be reviewed as-is, but I want to note that I'm also separately working on a set of tests that will make it easier to observe the actual changes to coverage mapping data (if any) caused by this PR.

So if that ends up being ready soon enough, it might make sense to defer merging PR this until after those tests are merged.

jackh726 commented:
@Zalathar I'll wait to review this for a bit. Let me know if the other PR takes more time than you want, or you'd rather I just review this.

Zalathar commented:
@jackh726 Yeah, currently the other change is proceeding well, so I think it makes sense to wait on reviewing this one. I'll speak up when that changes.

Zalathar commented:
@rustbot author

Zalathar committed:
coverage: Explicitly simplify coverage expressions in codegen

After coverage instrumentation and MIR transformations, we can sometimes end up
with coverage expressions that always have a value of zero. Any expression
operand that refers to an always-zero expression can be replaced with a literal
`Operand::Zero`, making the emitted coverage mapping data smaller and simpler.

This simplification step is mostly redundant with the simplifications performed
inline in `expressions_with_regions`, except that it does a slightly more
thorough job in some cases (because it checks for always-zero expressions
*after* other simplifications).

However, adding this simplification step will then let us greatly simplify that
code, without affecting the quality of the emitted coverage maps.

Zalathar commented:
I've rebased this to sit on top of a copy of #114843, which adds the new tests that demonstrate how this affects coverage maps. Once that gets merged and I rebase this onto the new master, the patches with 🍒 should all disappear.

Using those tests, I've found that there *are* changes to the maps, but they can be eliminated by adding an explicit simplify-expressions step during coverage codegen that replicates the simplifications performed in-passing by the renumbering code. (In fact, it does a slightly *better* job in some cases.)

Zalathar commented:
(Rebased onto a fresh copy of #114843 due to changes in that branch; no other changes.)

Zalathar commented:
@rustbot blocked on #114843

Zalathar commented:
^^ I think rustbot got very confused for some reason; the above pings are false positives and should be ignored.

Zalathar commented:
@rustbot label -A-meta -A-query-system -O-unix -O-windows -T-style

bors commented:
:umbrella: The latest upstream changes (presumably #115183) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #114843) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Zalathar commented:
@jackh726 The test suite has been merged (#114656), so this PR is now ready for review.

@rustbot ready
@rustbot label -A-testsuite -T-bootstrap

Zalathar commented:
(I'll keep rebasing occasionally to keep the history simple and to check that there are no soft conflicts.)

jackh726 reviewed (approved):
Very nice comments, made the review relatively easy.

jackh726 commented:
@bors r+

bors commented:
:pushpin: Commit 041a232b62dba833a4cfdca3f74e4029b53200a0 has been approved by `jackh726`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jackh726 041a232b62dba833a4cfdca3f74e4029b53200a0 -->
<!-- homu: {"type":"Approved","sha":"041a232b62dba833a4cfdca3f74e4029b53200a0","approver":"jackh726","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 041a232b62dba833a4cfdca3f74e4029b53200a0 with merge 24b45c3bdcd7435d57d222d1d47cb4ab2cf114d1...
<!-- homu: {"type":"BuildStarted","head_sha":"041a232b62dba833a4cfdca3f74e4029b53200a0","merge_sha":"24b45c3bdcd7435d57d222d1d47cb4ab2cf114d1"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6265069556/job/17016565098)
Approved by: jackh726
Pushing 24b45c3bdcd7435d57d222d1d47cb4ab2cf114d1 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"jackh726","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6265069556/job/17016565098"},"merge_sha":"24b45c3bdcd7435d57d222d1d47cb4ab2cf114d1"} -->

rust-timer commented:
Finished benchmarking commit (24b45c3bdcd7435d57d222d1d47cb4ab2cf114d1): [comparison URL](https://perf.rust-lang.org/compare.html?start=0fd7ce99b0508aff7f7a2c639871de4e8080e3f8&end=24b45c3bdcd7435d57d222d1d47cb4ab2cf114d1&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=0fd7ce99b0508aff7f7a2c639871de4e8080e3f8&end=24b45c3bdcd7435d57d222d1d47cb4ab2cf114d1&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.4% | [0.4%, 0.4%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=0fd7ce99b0508aff7f7a2c639871de4e8080e3f8&end=24b45c3bdcd7435d57d222d1d47cb4ab2cf114d1&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.5% | [-2.0%, -1.0%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -1.5% | [-2.0%, -1.0%] | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=0fd7ce99b0508aff7f7a2c639871de4e8080e3f8&end=24b45c3bdcd7435d57d222d1d47cb4ab2cf114d1&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=0fd7ce99b0508aff7f7a2c639871de4e8080e3f8&end=24b45c3bdcd7435d57d222d1d47cb4ab2cf114d1&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 630.615s -> 633.468s (0.45%)
**Artifact size**: 317.63 MiB -> 317.47 MiB (-0.05%)
